cli_reference/extend_cli.adoc

```diff
@@ -0,0 +1,218 @@
+
+. One or more directories specified according to the
+link:https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html[XDG
+System Directory Structure] specification.
++
+Specifically, the loader locates the directories specified by the
+`XDG_DATA_DIRS` environment variable, and then searches the *_kubectl/plugins_*
+directory inside of those. If `XDG_DATA_DIRS` is not specified, it defaults to
+*_/usr/local/share:/usr/share_*.
+
+. The `plugins` directory under the user's *_kubeconfig_* directory. In most
+cases, this is *_~/.kube/plugins_*:
++
+----
+# Loads plugins from both /path/to/dir1 and /path/to/dir2
+$ KUBECTL_PLUGINS_PATH=/path/to/dir1:/path/to/dir2 kubectl plugin -h
+----
+
+[[cli-writing-plugins]]
+== Writing Plug-ins
+
+You can write a plug-in in any programming language or script that allows you to
+write command-line commands. A plug-in does not necessarily need to have a
+binary component. It could rely entirely on operating system utilities like
+`echo`, `sed`, or `grep`. Alternatively, it could rely on the `oc` binary.
+
+The only strong requirement for an `oc` plug-in is the *_plugin.yaml_*
+descriptor file. This file is responsible for declaring at least the minimum
+attributes required to register a plug-in and must be located under one of the
+locations specified in the xref:cli-plugin-loader-search-order[Search Order]
+section.
+
+[[cli-writing-plugins-descriptor]]
+=== The plugin.yaml Descriptor
+
+The descriptor file supports the following attributes:
+
+----
+name: "targaryen"                 # REQUIRED: the plugin command name, to be invoked under 'kubectl'
+shortDesc: "Dragonized plugin"    # REQUIRED: the command short description, for help
+longDesc: ""                      # the command long description, for help
+example: ""                       # command example(s), for help
+command: "./dracarys"             # REQUIRED: the command, binary, or script to invoke when running the plugin
+flags:                            # flags supported by the plugin
+  - name: "heat"                  # REQUIRED for each flag: flag name
+    shorthand: "h"                # short version of the flag name
+    desc: "Fire heat"             # REQUIRED for each flag: flag description
+    defValue: "extreme"           # default value of the flag
+tree:                             # allows the declaration of subcommands
+  - ...                           # subcommands support the same set of attributes
+----
+
+The preceding descriptor declares the `oc plugin targaryen` plug-in, which has
+one flag named `-h | --heat`.
+
+When the plug-in is invoked, it calls the `dracarys` binary or script, which is
+located in the same directory as the descriptor file, passing a number of
+arguments and environment variables. The
+xref:cli-writing-plugins-accessing-runtime-attributes[Accessing Runtime
+Attributes] section describes how the `dracarys` command accesses the flag value
+and other runtime context.
```

Comment:
I hate to say it, but there might not be a place for GoT references in the docs... Can these be changed to something more generic? There's a few in the following paras as well.

Reply:
Ha! The references were lost on me because I am one of the few people in the world who doesn't watch GoT 🏰 🐲 Oops! yeah, we'll definitely want to clean those out. (Sorry @fabianofranz )

Reply:
but but but...  :)